src/distributions/semeru/installer.ts

```diff
@@ -0,0 +1,175 @@
+      .filter(item => isVersionSatisfies(version, item.version))
+      .sort((a, b) => {
+        return -semver.compareBuild(a.version, b.version);
+      });
+
+    const resolvedFullVersion = satisfiedVersions.length > 0 ? satisfiedVersions[0] : null;
+    if (!resolvedFullVersion) {
+      const availableOptions = availableVersionsWithBinaries.map(item => item.version).join(', ');
+      const availableOptionsMessage = availableOptions
+        ? `\nAvailable versions: ${availableOptions}`
+        : '';
+      throw new Error(
+        `Could not find satisfied version for SemVer '${version}'. ${availableOptionsMessage}`
+      );
+    }
+
+    return resolvedFullVersion;
+  }
+
+  protected async downloadTool(javaRelease: JavaDownloadRelease): Promise<JavaInstallerResults> {
+    let javaPath: string;
+    let extractedJavaPath: string;
+
+    core.info(
+      `Downloading Java ${javaRelease.version} (${this.distribution}) from ${javaRelease.url} ...`
+    );
+    const javaArchivePath = await tc.downloadTool(javaRelease.url);
+
+    core.info(`Extracting Java archive...`);
+    let extension = getDownloadArchiveExtension();
+
+    extractedJavaPath = await extractJdkFile(javaArchivePath, extension);
+
+    const archiveName = fs.readdirSync(extractedJavaPath)[0];
+    const archivePath = path.join(extractedJavaPath, archiveName);
+    const version = this.getToolcacheVersionName(javaRelease.version);
+
+    javaPath = await tc.cacheDir(archivePath, this.toolcacheFolderName, version, this.architecture);
+
+    return { version: javaRelease.version, path: javaPath };
+  }
+
+  protected get toolcacheFolderName(): string {
+    return super.toolcacheFolderName;
+  }
+
+  public async getAvailableVersions(): Promise<ISemeruAvailableVersions[]> {
+    const platform = this.getPlatformOption();
+    const arch = this.architecture;
+    const imageType = this.packageType;
+    const versionRange = encodeURI('[1.0,100.0]'); // retrieve all available versions
+    const releaseType = this.stable ? 'ga' : 'ea';
+
+    console.time('semeru-retrieve-available-versions');
+
+    const baseRequestArguments = [
+      `project=jdk`,
+      'vendor=ibm',
+      `heap_size=normal`,
+      'sort_method=DEFAULT',
+      'sort_order=DESC',
+      `os=${platform}`,
+      `architecture=${arch}`,
+      `image_type=${imageType}`,
+      `release_type=${releaseType}`,
+      `jvm_impl=openj9`
+    ].join('&');
+
+    // need to iterate through all pages to retrieve the list of all versions
+    // Adoptium API doesn't provide way to retrieve the count of pages to iterate so infinity loop
+    let page_index = 0;
+    const availableVersions: ISemeruAvailableVersions[] = [];
+    while (true) {
+      const requestArguments = `${baseRequestArguments}&page_size=20&page=${page_index}`;
+      const availableVersionsUrl = `https://api.adoptopenjdk.net/v3/assets/version/${versionRange}?${requestArguments}`;
```

Comment:
https://api.adoptopenjdk.net/ has been deprecated. Things have moved over to https://api.adoptium.net/ .

However, Eclipse/Adoptium doesn't (can't) distribute OpenJ9 builds.

Why don't you use https://github.com/foojayio/discoapi/ ? They have `semeru` and `semeru_certified` (distributions by IBM of OpenJ9; the `_certified` went through the Oracle's TCK test) available.

Reply:
I would love a stable API, but as it looks like now the one used by discoapi is unstandardised: https://github.com/foojayio/discoapi/blob/main/src/main/java/io/foojay/api/distribution/Semeru.java , if the setup-java action breaks because of AOJDK that's 'sad', but if the undocumented IBM marketplace changes their API and something breaks I do not want to be behind that... Unless it is documented and I am blind.

Also https://api.adoptopenjdk.net/;
> Clients that require OpenJ9/Semeru builds for now can continue to use AdoptOpenJDK. However, should look out for the upcoming Semeru api that will be the future delivery channel for Semeru builds.

And I do not have the time to reimplement the one used by discoapi anyway. I think for now we just need to get this integrated so people can just upgrade/get the new minor/patch version when introducing a future update to the retrieval mechanism, possibly when IBM adds a documented API.

Reply:
The Disco API documentation is here: https://api.foojay.io/swagger-ui
And the API is stable AFAIK.
I've only skimmed the Disco API code, but it seems that the packages are downloaded from GitHub (e.g. https://github.com/ibmruntimes/semeru17-binaries/releases ), which has a stable API.

It's only a matter of time when adoptopenjdk.net stops updating its index with newest versions of Semeru (and other distros), including versions with vulnerability patches, so this also has security implications.

That's why I thought that the actively developed and maintained Disco API is a safer bet than the deprecated adoptopenjdk.net for OpenJ9/Semeru users.